### PR TITLE
Feature/js api 2 9 0

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: assets/dev/js/editor/document/commands/base/history/debounce.js [hunk 1]
```diff
@@ -45,11 +47,38 @@ export default class Debounce extends History {
 		}
 	}
 
+	addTransaction() {
+		if ( Debounce.isTransactionStarted ) {
```
Comment (matipojo) [inline assets/dev/js/editor/document/commands/base/history/debounce.js]: It's  a bug fix?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/document/commands/base/history/debounce.js (hunk 1 above)
Comment (iNewLegend) [reply assets/dev/js/editor/document/commands/base/history/debounce.js]: no, just better way of interacting with debounce based commands. provide us way to free some objects:
```
		if ( Debounce.isTransactionStarted ) {
			delete this.history.title;
			delete this.history.subTitle;
		}
```

File: assets/dev/js/editor/document/history/component.js [hunk 2]
```diff
@@ -79,4 +78,8 @@ export default class Component extends ComponentBase {
 
 		return result;
 	}
+
+	isTransactionStarted() {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply assets/dev/js/editor/document/history/component.js]: There is no difference, the problem is that:

History/Component know when transaction is started and its logical... but
'debounce' need to know about transaction status and its un-logical to call History/Component from debounce-base...

**Review by matipojo — DISMISSED** [review]

**Review by matipojo — DISMISSED** [review]

File: assets/dev/js/editor/document/commands/base/history/debounce.js (hunk 1 above)
Comment (matipojo) [inline assets/dev/js/editor/document/commands/base/history/debounce.js]: Can it be moved to the the command? ('document/history/add-transaction')

File: assets/dev/js/editor/document/commands/base/history/debounce.js [hunk 3]
```diff
@@ -45,11 +47,38 @@ export default class Debounce extends History {
...
+
+		if ( ! Debounce.isTransactionStarted ) {
+			Debounce.isTransactionStarted = true;
+		}
+	}
+
+	deleteTransaction() {
+		$e.internal( 'document/history/delete-transaction' );
+
+		Debounce.isTransactionStarted = false;
```
Comment (matipojo) [inline assets/dev/js/editor/document/commands/base/history/debounce.js]: Can it be moved to the the command? ('document/history/delete-transaction')

File: assets/dev/js/editor/document/commands/base/history/debounce.js [hunk 4]
```diff
@@ -45,11 +47,38 @@ export default class Debounce extends History {
...
+	}
+
+	endTransaction() {
+		if ( Debounce.isTransactionStarted ) {
```
Comment (matipojo) [inline assets/dev/js/editor/document/commands/base/history/debounce.js]: Can it be moved to the the command? ('document/history/end-transaction')

File: assets/dev/js/editor/document/helper-bc.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+export default class BackwardsCompatibility {
```
Comment (matipojo) [inline assets/dev/js/editor/document/helper-bc.js]: Add them to the Deprecation List file

File: core/common/assets/js/api/modules/component-base.js [hunk 6]
```diff
@@ -3,12 +3,19 @@ export default class ComponentBase extends elementorModules.Module {
 		if ( args.manager ) {
 			this.manager = args.manager;
 		}
+		/**
```
Comment (matipojo) [inline core/common/assets/js/api/modules/component-base.js]: Move to a task

File: core/common/assets/js/api/modules/component-base.js [hunk 7]
```diff
@@ -22,6 +29,8 @@ export default class ComponentBase extends elementorModules.Module {
 		jQuery.each( this.getCommands(), ( command, callback ) => this.registerCommand( command, callback ) );
 
 		jQuery.each( this.getCommandsInternal(), ( command, callback ) => this.registerCommandInternal( command, callback ) );
+
+		Object.values( this.defaultHooks() ).forEach( ( hook ) => this.registerHook( hook ) );
```
Comment (matipojo) [inline core/common/assets/js/api/modules/component-base.js]: Should be like the others. (ES6/jQuery)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/document/history/commands/internal/add-transaction.js [hunk 8]
```diff
@@ -1,6 +1,20 @@
-import Base from './base/base';
+import Base from '../base/base';
 
 export class AddTransaction extends Base {
+	initialize( args ) {
+		super.initialize( args );
+
+		if ( this.component.isTransactionStarted() ) {
```
Comment (matipojo) [inline assets/dev/js/editor/document/history/commands/internal/add-transaction.js]: Explain why.

File: core/common/assets/js/api/modules/component-base.js [hunk 9]
```diff
@@ -6,22 +6,26 @@ export default class ComponentBase extends elementorModules.Module {
...
-		jQuery.each( this.getRoutes(), ( route, callback ) => this.registerRoute( route, callback ) );
+		Object.entries( this.getRoutes() ).forEach( ( [ route, callback ] ) => this.registerRoute( route, callback ) );
 
-		jQuery.each( this.getCommands(), ( command, callback ) => this.registerCommand( command, callback ) );
+		Object.entries( this.getCommands() ).forEach( ( [ command, callback ] ) => this.registerCommand( command, callback ) );
 
-		jQuery.each( this.getCommandsInternal(), ( command, callback ) => this.registerCommandInternal( command, callback ) );
+		Object.entries( this.getCommandsInternal() ).forEach( ( [ command, callback ] ) => this.registerCommandInternal( command, callback ) );
+
+		Object.values( this.defaultHooks() ).forEach( ( hook ) => this.registerHook( hook ) );
```
Comment (matipojo) [inline core/common/assets/js/api/modules/component-base.js]: Why not `Object.entries`?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/document/commands/base/history/debounce.js (hunk 1 above)
Comment (matipojo) [reply assets/dev/js/editor/document/commands/base/history/debounce.js]: Where is the code now?

File: assets/dev/js/editor/regions/panel/pages/editor/component.js [hunk 10]
```diff
@@ -24,10 +24,10 @@ export default class Component extends ComponentBase {
 	defaultCommands() {
 		return {
 			open: ( args ) => {
-				this.openEditor( args.model, args.view );
```
Comment (matipojo) [inline assets/dev/js/editor/regions/panel/pages/editor/component.js]: Try revert the order.

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — DISMISSED** [review]

File: modules/history/assets/js/history/panel-tab.js [hunk 11]
```diff
@@ -55,9 +55,9 @@ module.exports = Marionette.CompositeView.extend( {
 			return;
 		}
 
-		var collection = event.model.collection,
-			itemIndex = collection.findIndex( event.model );
+		const collection = event.model.collection,
+			index = collection.findIndex( event.model );
 
-		this.options.history.doItem( itemIndex );
+		$e.run( 'panel/history/actions/do', { index } );
```
Comment (matipojo) [inline modules/history/assets/js/history/panel-tab.js]: internal

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — APPROVED** [review]

File: assets/dev/js/editor/elements/models/element.js [hunk 12]
```diff
@@ -7,7 +7,9 @@ ElementModel = Backbone.Model.extend( {
 		elType: '',
 		isInner: false,
 		settings: {},
-		defaultEditSettings: {},
+		defaultEditSettings: {
+			editTab: 'content',
```
Comment (kobizz) [inline assets/dev/js/editor/elements/models/element.js]: ?

File: assets/dev/js/editor/elements/views/widget.js [hunk 13]
```diff
@@ -110,6 +110,7 @@ WidgetView = BaseElementView.extend( {
 					name: 'save',
 					title: elementor.translate( 'save_as_global' ),
 					shortcut: jQuery( '<i>', { class: 'eicon-pro-icon' } ),
+					isEnabled: () => 'global' !== this.options.model.get( 'widgetType' ),
```
Comment (kobizz) [inline assets/dev/js/editor/elements/views/widget.js]: ?

File: assets/dev/js/editor/elements/models/element.js (hunk 12 above)
Comment (iNewLegend) [reply assets/dev/js/editor/elements/models/element.js]: ?

File: assets/dev/js/editor/elements/views/widget.js (hunk 13 above)
Comment (iNewLegend) [reply assets/dev/js/editor/elements/views/widget.js]: ?

**Pull request merged**